extmod/modbluetooth: Allow config of scan interval/window.

jimmo opened the pull request:
This adds two additional optional kwargs to `gap_scan()`:
  - `interval_ms`: How long between scans.
  - `window_ms`: How long to scan for during a scan.

The default with NimBLE is a 11.25ms window with a 1.28s interval.

Changing these parameters is important for detecting low-frequency
advertisements (e.g. beacons).

jimmo committed:
extmod/modbluetooth: Allow config of scan interval/window.

This adds two additional optional kwargs to `gap_scan()`:
  - `interval_us`: How long between scans.
  - `window_us`: How long to scan for during a scan.

The default with NimBLE is a 11.25ms window with a 1.28s interval.

Changing these parameters is important for detecting low-frequency
advertisements (e.g. beacons).

Note: these params are in microseconds, not milliseconds in order
to allow the 625us granularity offered by the spec.

dpgeorge commented on extmod/modbluetooth.c:
```diff
@@ -500,24 +500,26 @@ STATIC mp_obj_t bluetooth_ble_gap_connect(size_t n_args, const mp_obj_t *args) {
+        { MP_QSTR_window_ms, MP_ARG_INT, {.u_int = 0} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
```
Instead of kwargs it can just use positional with defaults.  That should save code size, no need for qstrs nor the table of allowed args.

And I think the default values should be specified, eg to be what they currently are.  Something like:
```python
mp_int_t interval_ms = 1280;
mp_int_t window_ms = 11;
if (n_args > 2) {
    interval_ms = mp_obj_get_int(args[2]);
    if (n_args > 3) {
        window_ms = mp_obj_get_int(args[3]);
    }
}
```

dpgeorge replied:
Oh, I see that `gap_advertise` uses kwargs.  Nevertheless I still think it makes sense to keep `gap_scan` "simple" and just use positional here.

dpgeorge commented on extmod/modbluetooth_nimble.c:
```diff
@@ -614,16 +614,24 @@ STATIC int gap_scan_cb(struct ble_gap_event *event, void *arg) {
-        .window = BLE_GAP_SCAN_SLOW_WINDOW1,
+    uint16_t interval = BLE_GAP_SCAN_SLOW_INTERVAL1;
+    uint16_t window = BLE_GAP_SCAN_SLOW_WINDOW1;
+    if (interval_ms > 0) {
```
No need to check this, just assume valid (default at least) values are passed in.

dpgeorge commented on extmod/modbluetooth_nimble.c:
```diff
@@ -614,16 +614,24 @@ STATIC int gap_scan_cb(struct ble_gap_event *event, void *arg) {
         .filter_policy = BLE_HCI_CONN_FILT_NO_WL,
         .limited = 0,
-        .passive = 0,
+        .passive = 1,  // TODO: Handle BLE_HCI_ADV_RPT_EVTYPE_SCAN_RSP in gap_scan_cb above.
```
Why does this need to be changed?

jimmo replied:
It was wrong before, because we're not handling the scan response event. Now that this change allows us to scan much faster, no point spamming the radio with active scans if we're going to ignore them.

I plan to add an `active=True` arg to scan() later when implementing active scans.

jimmo commented:
Thanks, have made the suggested changes.

dpgeorge commented:
Looking at this again I see that the scanning intervals are in millisecond units.  But this doesn't give enough resolution (in integers) to match what's possible with BLE.  BLE scanning (and advertising) uses units of 625us, and so allows fractional millisecond values.

So I think it'd be good to change the units of these interval parameters to **microseconds**.  (But keep the scan duration in milliseconds, which is not related to the BLE spec.)

jimmo committed:
extmod/modbluetooth: Use us instead of ms for advertising interval.

This is to more accurately match the BLE spec, where intervals are
configured in units of channel hop time (625us). When it was
specified in ms, not all "valid" intervals were able to be
specified.

Now that we're also allowing configuration of scan interval, this
commit updates advertising to match.
